### DISCUSSION TITLE
deform, distort, transform, rectangular image to an ellipse

### DISCUSSION THREAD

**Discussion by engilic** [ImageMagick/ImageMagick#6153] [Digital Image Processing]
Hi, I'm new here,

Tried this with ChatGPT but with no success.

How to deform, distort, transform, or squeeze rectangular photos to an ellipse, similar to Adobe Illustrator Envelope Distort, but better.
Want to see how the photo will look if it is squeezed into an ellipse so all points go to the center respecting the distance from the center?
Every point is moved toward the center by the ratio of the Initial distance / distance of the ellipse radius at the point where the line from the point that is moving intersects the ellipse.
The edge of the rectangle picture sits exactly on the ellipse.

Something like squeezing a rectangle to ellipse from the edge to the center evenly respecting the distance from the center.

Tried to explain as best as I could.

Thank you

**Comment by fmw42** [comment]
That requires a conformal mapping. See https://arxiv.org/pdf/1509.06344.pdf

**Reply by engilic** [reply]
this is great, thank you

Any ideas on how to apply those formulas to distort a rectangle?

**Comment by fmw42** [comment]
See https://imagemagick.org/Usage/mapping/

Use -fx to do the warping from the equations given for mapping between square and circle.  I don't have references for rectangle and ellipse.

**Comment by snibgo** [comment]
This does what I think you want. Windows BAT script.
```
set rx=250
set ry=200

%IMG7%magick ^
  %1 ^
  -resize 200%% ^
  -set option:MYSIZE %%[fx:w]x%%[fx:h] ^
  -virtual-pixel Edge ^
  -distort DePolar -1,0 ^
  ( -size %%[MYSIZE] ^
    -define gradient:radii=%rx%,%ry% ^
    radial-gradient:Black-White ^
    -background White -virtual-pixel Background ^
    -distort DePolar -1,0 ^
  ) ^
  ( -size %%[MYSIZE] xc:White ^
    -virtual-pixel Black ^
    -distort DePolar -1,0 ^
    -set option:MYSIZE2 %%[fx:w]x%%[fx:h] ^
    -scale "x1^!" ^
    -scale "%%[MYSIZE2]^!" ^
  ) ^
  ( -clone 1-2 ^
    -compose Multiply -composite ^
    -channel R -fx "i/(w-1)" +channel ^
  ) ^
  -delete 1-2 ^
  -compose Distort -set option:compose:args 100%%x100%% -composite ^
  -virtual-pixel Edge ^
  -distort Polar -1,0 ^
  ( -size %%[MYSIZE] xc:Black ^
    -fill White -stroke None ^
    -draw "ellipse %%[fx:w/2],%%[fx:h/2],%rx%,%ry%,0,360" ^
  ) ^
  -alpha off ^
  -compose CopyOpacity -composite ^
  -resize 50%% ^
  %2
```
Note that this supersamples for quality. The ellipse radii are in pixels, and apply to the supersampled image.

Input toes_grid.jpg:

![toes_grid](https://user-images.githubusercontent.com/33812368/224560562-af05eaf8-2948-49e1-9e86-5acaeb01990c.jpg)

Output:

![r5](https://user-images.githubusercontent.com/33812368/224560572-a0837371-9804-420b-99a0-b00f1f9d9fd3.jpg)

**Reply by engilic** [reply]
This is great.
I read the article above and the simple transformation gives result which looks like it is broken, and the formulas are not for rectangle and ellipse, they are for square and circle, and adjusting them you need to play with radiuses and the difference between circle and ellipse formula, which is not easy.

**Comment by fmw42** [comment]
Excellent!  But for others, you might explain what your major sections do or just your concept. Or I suppose they could put +write statements at various places.

**Comment by snibgo** [comment]
The work is done on the image "unrolled" by `-distort DePolar`, so the x-axis is what was the angle around the centre, and the y-axis is what was the distance from the centre. The distortion is then just in the y-direction. At the end we "roll up" the image by `-distort Polar`.

The first image in parentheses is an elliptical gradient, which is unrolled so the result is black at the top and white at the edge of the ellipse. The next image in parentheses is a white version of the image, unrolled, and scaled to one pixel high so each pixel is the proportion of the unrolled image that was in the original rectangle.

Multiplying these together makes an absolute displacement map in the y-direction, for `-compose Distort ...`.

The result contains virtual pixels. I make them transparent by copying opacity from a a drawn ellipse.

The multiplication and distortion could instead be done in a single `-fx` operation. Then we also wouldn't need the `-scale "%%[MYSIZE2]^!"`. This would simplify the code, and probably increase performance.

**Reply by engilic** [reply]
Is there any way to make this to work like Elliptical Grid Mapping, its still not exactly what I wanted because its deforming edges much more than the centre part, looks like an reflection from a ball. That is exactly what Illustrator Envelope Distort does.

what you did is extraordinary and I know asking more is ridiculous but I cannot use what you made because as I said its braking the lines, yes it is what i wanted but now I se it does not give the result I wanted. We say “be careful what you wish, it might come true”
If you have time snd will it would be nice, if not, you did a great job and someone might use it

THANK YOU VERY MUCH

**Comment by snibgo** [comment]
The two papers by Fong show a variety of methods that can easily be implemented with IM. For example, the "Simple Stretching" method:
```
set rx=250
set ry=200

set /A dx=%rx%*2
set /A dy=%ry%*2

set sFX=^
uu=(i+0.5)/%dx%-0.5; ^
vv=(j+0.5)/%dy%-0.5; ^
hyp=hypot(uu,vv); ^
if (uu*uu ^>= vv*vv, ^
    xx=sign(uu)*hyp; yy=sign(uu)*vv/uu*hyp; , ^
    xx=sign(vv)*uu/vv*hyp; yy=sign(vv)*hyp; ^
   ) ; ^
p{ (xx+0.5)*w-0.5, (yy+0.5)*h-0.5 }

%IMG7%magick ^
  toes_grid.jpg ^
  -resize 200%% ^
  -virtual-pixel None ^
  -fx "%sFX%" ^
  -resize 50%% ^
  x2.png
```

The IM command is much simpler than my command above, but with a complex `-fx` operation. I have written the `-fx` expression as a separate environment variable because I can then include indentations etc for understandability. The result is:

![x2](https://user-images.githubusercontent.com/33812368/224736180-27b88740-41e3-4b9c-af10-541dd5672c9b.png)

When the ellipse and rectangle have the same aspect ratios, the result is the same as my complex command above. When the aspect ratios are different, the Fong method doesn't retain the angles from the centre, so pixels are not simply moved along a radius towards the centre. But it might do what you want, of course.

> ... but now I see it does not give the result I wanted.

Fair enough. Can you explain what result you do want?

**Comment by snibgo** [comment]
Here is Fong's "Elliptical Grid Mapping":
```
set rx=250
set ry=200

set /A dx=%rx%*2
set /A dy=%ry%*2

set sFX=^
uu=2*((i+0.5)/%dx%-0.5); ^
vv=2*((j+0.5)/%dy%-0.5); ^
twosq=%%[fx:2*sqrt(2)]; ^
usq=uu*uu; ^
vsq=vv*vv; ^
xx=0.5 * ( sqrt(2+usq-vsq+twosq*uu) - sqrt(2+usq-vsq-twosq*uu) ); ^
yy=0.5 * ( sqrt(2-usq+vsq+twosq*vv) - sqrt(2-usq+vsq-twosq*vv) ); ^
p{ (xx/2+0.5)*w-0.5, (yy/2+0.5)*h-0.5 }

%IMG7%magick ^
  toes_grid.jpg ^
  -resize 200%% ^
  -virtual-pixel None ^
  -fx "%sFX%" ^
  -resize 50%% ^
  x3.png
```

![x3](https://user-images.githubusercontent.com/33812368/224785420-472b89ce-3d32-4267-a815-7ba53b192af2.png)

**Comment by engilic** [comment]
Thank you
